tests/test_cases/test_cocotb/test_concurrency_primitives.py

```diff
@@ -165,3 +168,55 @@ async def coro(delay):
     test_start = cocotb.utils.get_sim_time(units="ns")
     await Combine(*coros)
     assert cocotb.utils.get_sim_time(units="ns") == test_start + max_delay
+
+
+@cocotb.test()
+async def test_recursive_combine_and_fork(_):
+    """ Test using `Combine` on forked coroutines that themselves use `Combine`.
+
+    This does not test passing `Combine` triggers into `Combine`. """
+
+    async def mergesort(n):
+        if len(n) == 1:
+            return n
+        part1 = n[:len(n) // 2]
+        part2 = n[len(n) // 2:]
+        sort1 = cocotb.fork(mergesort(part1))
+        sort2 = cocotb.fork(mergesort(part2))
+        await Combine(sort1, sort2)
+        res1 = deque(sort1.retval)
+        res2 = deque(sort2.retval)
+        res = []
+        while res1 and res2:
+            if res1[0] < res2[0]:
+                res.append(res1.popleft())
+            else:
+                res.append(res2.popleft())
+        res.extend(res1)
+        res.extend(res2)
+        return res
+
+    t = [randint(0, 1000) for _ in range(100)]
+    res = await mergesort(t)
+    t.sort()
+    assert t == res
+
+
+@cocotb.test()
+async def test_recursive_combine(_):
+    """Tests using a `Combine` trigger in another `Combine` trigger"""
+
+    async def waiter(N):
+        await Timer(N, 'ns')
+
+    start_time = get_sim_time('ns')
+    await Combine(
+        Combine(
+            cocotb.fork(waiter(10)),
+            cocotb.fork(waiter(20))
+        ),
+        cocotb.fork(waiter(30))
+    )
+    end_time = get_sim_time('ns')
+
+    assert end_time - start_time == 30
```

Comment:
It would be good to verify that all three coroutines run, perhaps something like (untested):
```suggestion
    done = set()
    async def waiter(N):
        await Timer(N, 'ns')
        done.add(N)

    start_time = get_sim_time('ns')
    await Combine(
        Combine(
            cocotb.fork(waiter(10)),
            cocotb.fork(waiter(20))
        ),
        cocotb.fork(waiter(30))
    )
    end_time = get_sim_time('ns')

    assert end_time - start_time == 30
    assert done == {10, 20, 30}
```